Add __experimentalUseResourcePermissions

adamziel opened the pull request:
## Description
This PR is a minimal subset of https://github.com/WordPress/gutenberg/pull/38135 focused solely on the `useResourcePermissions` hook.

The goal of these new APIs is to lower the barrier of entry for new contributors and make the life of existing contributors easier.

### Example usage:
**Before**
```js
useSelect(
  ( select ) => {
	  return {
		canUserUpdateNavigationEntity: ref
			? canUser( 'update', 'navigation', ref )
			: undefined,
		hasResolvedCanUserUpdateNavigationEntity: hasFinishedResolution(
			'canUser',
			[ 'update', 'navigation', ref ]
		),
		canUserDeleteNavigationEntity: ref
			? canUser( 'delete', 'navigation', ref )
			: undefined,
		hasResolvedCanUserDeleteNavigationEntity: hasFinishedResolution(
			'canUser',
			[ 'delete', 'navigation', ref ]
		),
		canUserCreateNavigation: canUser( 'create', 'navigation' ),
		hasResolvedCanUserCreateNavigation: hasFinishedResolution(
			'canUser',
			[ 'create', 'navigation' ]
		),
	  };
  },
  [ ref ]
);
```

**After:**
```js
const [ hasResolved, { canCreate, canUpdate, canDelete } ] = __experimentalUseResourcePermissions( 'navigation', ref );
```

It returns a tuple and with `hasResolved` the first item so that it's intentionally too hard to overlook. Thank you @kevin940726 for this suggestion!

See also [how the combination of all proposed hooks makes the navigation block ~200 lines leaner](https://github.com/WordPress/gutenberg/pull/38289).

### Test plan

1. Confirm if the tests are green
2. Confirm if the code makes sense

### Other considerations

@gziolo [said](https://github.com/WordPress/gutenberg/pull/38135/files#r794378548):
> Nit: for consistency, it might use a more explicit name like useEntityRecordPermissions since it works only for a single record.

API resources are separate from entity records, we refer to them using a different set of identifiers.

> What's the story for multiple records?

You can either pass an id to this hook and get a create/update/delete permissions, or skip an id and get only the create permissions. Both are useful, e.g. in the Gutenberg codebase we sometimes do `canUser( 'update', 'settings' );` and other times we do `canUser( 'delete', resource, postId )`.

cc @kevin940726 @talldan @gziolo @draganescu @ellatrix @noisysocks @jsnajdr @getdave @Mamaduka @spencerfinnell

Mamaduka commented on packages/core-data/src/hooks/use-resource-permissions.ts:
```diff
@@ -0,0 +1,112 @@
+				status,
+				isResolving,
+				hasResolved,
+				canCreate: hasResolved && create.data,
+				canUpdate: hasResolved && update.data,
+				canDelete: hasResolved && _delete.data,
```
Just want to link @gziolo's suggestion from the original PR - https://github.com/WordPress/gutenberg/pull/38135#discussion_r794380681.

Mamaduka commented:
We can use this hook to replace [`useCanEditEntity`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/utils/hooks.js#L14-L20) and showcase usage in the core.

gziolo commented on packages/core-data/src/hooks/test/use-resource-permissions.js:
```diff
@@ -0,0 +1,109 @@
+import { store as coreDataStore } from '../../index';
+import useResourcePermissions from '../use-resource-permissions';
+
+describe( 'useEntityRecord', () => {
```
Nit: copy/paste issue

kevin940726 commented on packages/core-data/src/hooks/use-resource-permissions.ts:
```diff
@@ -0,0 +1,112 @@
+			} else if ( hasResolved ) {
+				status = Status.Success;
+			}
+			return {
```
I wonder if we should guard against incorrect use when `hasResolved` is not `true`. By returning a tuple and making `hasResolved` the first item of the tuple forces the users to think about that variable and it's not easy to accidentally miss it.

```js
const [ hasResolved, { canCreate } ] = useResourcePermissions( 'pages' );
```

One can still intentionally skip it but it immediately becomes weird and could trigger some additional care in code reviews.

adamziel replied:
I really like it @kevin940726, I just made that change 🙌

adamziel committed:
Squashed commit of the following:

commit c86f5616a89cb1998ead4ea1f34694795e019267
Author: Adam Zieliński <adam@adamziel.com>
Date:   Wed Feb 16 12:17:33 2022 +0100

    Move the return type to TS

commit eefec13ba88c7aa63e950b71dc242de4b9a46b16
Author: Adam Zieliński <adam@adamziel.com>
Date:   Wed Feb 16 12:16:36 2022 +0100

    Adjust TS types

commit 12c74338a1b63d8b796b50e11160bf89fc2f5c49
Author: Adam Zieliński <adam@adamziel.com>
Date:   Wed Feb 16 12:13:22 2022 +0100

    Make useResourcePermissions return a tuple and make hasResolved the first item of the tuple forces the users to think about that variable and it's not easy to accidentally miss it.

commit caa48a72800e6b47934a8cef5356b32bb8361788
Author: Adam Zieliński <adam@adamziel.com>
Date:   Tue Feb 15 17:03:31 2022 +0100

    Fix typo in the tests

commit 33cdb4e59fd3a81c743430fa6016903373aff666
Author: Adam Zieliński <adam@adamziel.com>
Date:   Mon Feb 14 15:05:09 2022 +0100

    Expose __experimentalUseResourcePermissions as a public API

commit cf30c5aa85dae7d7a488573b5e6459abc05086d5
Author: Adam Zieliński <adam@adamziel.com>
Date:   Mon Feb 14 14:58:42 2022 +0100

    Distinguish between global and local resoluion

commit 0db7bd25a7106400188addc23892d77bd33b63ed
Author: Adam Zieliński <adam@adamziel.com>
Date:   Mon Feb 14 14:42:36 2022 +0100

    Propose useResourcePermissions hook

commit 2d5e270d67781783c02d1d8ff1fa0d3a671e3aff
Author: Adam Zieliński <adam@adamziel.com>
Date:   Mon Feb 14 14:48:09 2022 +0100

    Move the status computation inside the enriched selectors

commit e7ac34e5e9fdb3051d9d093fad5d4ab73833ba0a
Author: Adam Zieliński <adam@adamziel.com>
Date:   Mon Feb 14 14:01:11 2022 +0100

    Propose useEntityRecords

kevin940726 commented on packages/core-data/src/hooks/use-resource-permissions.ts:
```diff
@@ -0,0 +1,120 @@
+ */
+export default function __experimentalUseResourcePermissions< IdType = void >(
+	resource: string,
+	id: IdType
```
I guess we should make it optional in type?

```suggestion
	id?: IdType
```

(Not sure if it's needed)

adamziel replied:
Good call!

adamziel committed:
Update packages/core-data/src/hooks/use-resource-permissions.ts

Co-authored-by: Kai Hao <kevin830726@gmail.com>

adamziel commented:
I'd like to update this PR once https://github.com/WordPress/gutenberg/pull/41139 lands

adamziel committed:
Migrate useNavigationMenu to useResourcePermissions

adamziel committed:
Fix unit tests

adamziel commented:
I fixed the unit tests, rebased this on top of trunk, and added a refactor of `useNavigationMenu` to showcase the new powers. @kevin940726 @getdave would you be up for reviewing?

kevin940726 reviewed (approved):
Unit tests are failing, but other than that I think this looks good 👍

adamziel committed:
Fix unit tests

adamziel committed:
Remove the http method check in navigation.test.js

adamziel committed:
Remove the testNavId from the mocked reponse endpoint in navigation.test.js

mburridge commented:
Added the Needs Dev Note label in case this needs a dev note (either individual or as part of a "misc" dev note) for WP 6.1 release.

adamziel commented:
Dev note proposed at https://make.wordpress.org/core/?p=99829&preview=1&_ppp=2b1db975ea
